### PR TITLE
Addition of Parquet Output Type

### PR DESCRIPTION
**Pull request by annarudy**
- Added --write_to_parquet argument flag that indicates output type should be parquet when included.
- Included beam step to write to parquet. 
- Created util to translated the json schema in schema_file to a pyarrow schema, which is a required input for beam.io.WriteToParquet(), along with unit test.

### CODE REVIEW

**Comment by jaketf** [comment]
@annarudy did you commit some of this from a different github account?

File: examples/dataflow-data-generator/data-generator-pipeline/data_generator/ParquetUtil.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            logging.error('Error: json schema included a {0:s} field. BYTE,  '
+                          'GEOGRAPHY, and RECORD types cannot be used when '
+                          'outputting to parquet.'.format(field_type))
+            raise ValueError('Error: json schema included a {0:s} field. BYTE, '
+                             'GEOGRAPHY, and RECORD types cannot be used when '
+                             'outputting to parquet.'.format(field_type))
+        else:
+            pa_field = pa.field(
+                field_name,
+                converted_field_type
```
Comment (jaketf) [inline examples/dataflow-data-generator/data-generator-pipeline/data_generator/ParquetUtil.py]: you need to check if the field is NULLABLE in the `string_schema` and set nullability with the kwargs `nullable=True` or `nullable=False`in the constructor for `pa_field`

File: examples/dataflow-data-generator/README.md [hunk 2]
```diff
@@ -136,6 +136,18 @@ Output format is specified by passing one of the `--csv_schema_order` or `--avro
 --avro_schema_file=/path/to/linorders.avsc
 ```
 
+`--write_to_parquet` is a flag that specifies the output should be parquet. In order for beam to write to parquet, 
+a pyarrow schema is needed. Therefore, this tool translates the schema in the --schema_file to
+a pyarrow schema automatically if this flag is included, but pyarrow doesn't support all fields that are supported
+by BigQuery. STRING, NUMERIC, INTEGER, FLOAT, NUMERIC, BOOLEAN, TIMESTAMP, DATE, TIME, and DATETIME types are supported. 
+However BYTE, GEOGRAPHY, and RECORD fields are not supported and cannot be included in the --schema_file when writing
```
[the remark this answers is not in the document]
Comment (jaketf) [reply examples/dataflow-data-generator/README.md]: The [docs](https://arrow.apache.org/docs/python/data.html#type-metadata) on how to use the `pyarrow.struct()` factory.

[the remark this answers is not in the document]
Comment (jaketf) [reply examples/dataflow-data-generator/README.md]: You should be able to handle this by making this schema converter recursive.

File: examples/dataflow-data-generator/data-generator-pipeline/data_generator_pipeline.py [hunk 3]
```diff
@@ -125,6 +125,19 @@ def run(argv=None):
                 )
         )
 
+    if data_args.write_to_parquet:
+        with open(data_args.schema_file, 'r') as infile:
+            str_schema = json.load(infile)
+        pa_schema = get_pyarrow_translated_schema(str_schema)
+        (rows
```
Comment (jaketf) [inline examples/dataflow-data-generator/data-generator-pipeline/data_generator_pipeline.py]: Have you done a test writing TIMESTAMP type fields? I'm curious if we need a date handler like [`fix_record_for_avro`](https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/dataflow-data-generator/data-generator-pipeline/data_generator/AvroUtil.py#L17) to cast from a date string to a unix timestamp. The data generator is going to create this field in "yyyy/MM/ddTHH:mm:SS" see [`sanity_check`](https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/dataflow-data-generator/data-generator-pipeline/data_generator/PrettyDataGenerator.py#L331)
https://arrow.apache.org/docs/python/generated/pyarrow.timestamp.html

File: examples/dataflow-data-generator/data-generator-pipeline/data_generator/ParquetUtil.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import pyarrow as pa
```
Comment (jaketf) [inline examples/dataflow-data-generator/data-generator-pipeline/data_generator/ParquetUtil.py]: Add Apache Header.

File: examples/dataflow-data-generator/data-generator-pipeline/data_generator_pipeline.py (hunk 3 above)
Comment (jaketf) [reply examples/dataflow-data-generator/data-generator-pipeline/data_generator_pipeline.py]: Alternatively, you could change how we generate dates upstream.

File: examples/dataflow-data-generator/data-generator-pipeline/tests/test_get_pyarrow_translated_schema.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import json
```
Comment (jaketf) [inline examples/dataflow-data-generator/data-generator-pipeline/tests/test_get_pyarrow_translated_schema.py]: Add Apache Header.

File: examples/dataflow-data-generator/data-generator-pipeline/data_generator/ParquetUtil.py (hunk 1 above)
Comment (jaketf) [reply examples/dataflow-data-generator/data-generator-pipeline/data_generator/ParquetUtil.py]: Additionally, you should handle creating an array if the input field mode is REPEATED.
You'll have to do this using the [`pyarrow.list_(value_type)`](https://arrow.apache.org/docs/python/generated/pyarrow.list_.html#pyarrow.list_) factory method.

File: examples/dataflow-data-generator/data-generator-pipeline/tests/test_get_pyarrow_translated_schema.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import logging
+import pyarrow as pa
+import unittest
+from data_generator.ParquetUtil import get_pyarrow_translated_schema
+
+
+class TestPyarrowSchemaTranslator(unittest.TestCase):
+    def test_get_pyarrow_translated_schema(self):
```
Comment (jaketf) [inline examples/dataflow-data-generator/data-generator-pipeline/tests/test_get_pyarrow_translated_schema.py]: Once you add the functionality for RECORD type, change this test schema to have a nested field.
Also, once you add date/ timestamp handling similar to [`fix_record_for_avro`](https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/dataflow-data-generator/data-generator-pipeline/data_generator/AvroUtil.py#L17) you should construct a record and test the date handling similar to [`test_fix_record_for_avro`](https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/dataflow-data-generator/data-generator-pipeline/tests/test_fix_record_for_avro.py)

File: examples/dataflow-data-generator/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaketf) [reply examples/dataflow-data-generator/README.md]: I just realized the PR for adding record types hasn't been merged. Which is probably why you chose not to support RECORD. If you could add the logic for parsing the schema that'd decrease the back and forth one that other PR is merged.

[the remark this answers is not in the document]
Comment (annarudy) [reply examples/dataflow-data-generator/README.md]: hey Jake, I actually had this comment in mind: https://docs.google.com/a/google.com/document/d/1SMEDTR2EaqaKIFKqN2rWgXMSkUO2gCU5CXxxKRDHoX4/edit?disco=AAAACkX0EUc . I guess that wasn't very clear in the readme. My thought was that even though pyarrow does support record types, parquet loads can't.

File: examples/dataflow-data-generator/data-generator-pipeline/data_generator_pipeline.py (hunk 3 above)
Comment (annarudy) [reply examples/dataflow-data-generator/data-generator-pipeline/data_generator_pipeline.py]: I can definitely look into this

**Comment by annarudy** [comment]
updating commits' authors doesn't seem to be working - going to try to close this and redo with the correct author

**Comment by annarudy** [comment]
Couldn't get rebase to fix failed CLA check here, so I duplicated branch parquetio_addition, created a new PR with the exact same changes (#202), and delete this PR.

**Pull request closed without merging**